### ISSUE TITLE
Make UI Classes Public and add method of presenting custom Subclasses

### ISSUE THREAD

**Issue by alexpopov** [FolioReader/FolioReaderKit#71]
I absolutely love the project, but I needed to make some fairly trivial changes to the UI but there is no way to do that out of the box. I had to fork the project, make a couple of the ViewControllers public and override a method just to add a bar button to the Navigation Bar. 

My proposal: making the main UI classes public so that they can be overridden, and add a method on `FolioReader` to present a custom `FolioReaderContainer`. 

I know this is not at all a trivial task: it would require stream-lining the internal API to make it fit for public consumption; eschewing some of the internal-state assumptions that the project currently hides, such as the implicitly unwrapped optional values; and adding comprehensive documentation, e.g. specifying where calls to `super` are required and what [pre | post]conditions must be met. 

I'd love to help in this, as your project helped me scrap a week of work building a barely-functional JS-based in-app reader, and instead perfectly integrate your work, along with my custom changes, in just a day. 

If, however, you want to maintain the project's simplicity and easy drop-in nature, I totally understand.

**Comment by hebertialmeida** [comment]
Hey @alexpopov I definitely want your help on this, I would love to make it more flexible and customisable.

**Comment by alexpopov** [comment]
I don't want to change the architecture too much, but my basic implementation proposal is to refactor some of the more monolithic `viewDidLoad` methods, and compose them of `setup*` and `create*` methods. `create*` methods handle our internal state (and should probably stay `internal`), `setup*` methods return what was previously instantiated inline. 
Here is an example from my rough mini-fork for work:
```swift
override func viewDidLoad() {
  super.viewDidLoad()
  createCenterViewController()
  // ... create more stuff
}

func createCenterViewController() {
  centerViewController = setupCenterViewController()
  // internal state stuff that we *need* to work goes here
  centerViewController.folioReaderContainer = self
  FolioReader.sharedInstance.readerCenter = centerViewController
  // ... 
}

public func setupCenterViewController() -> FolioReaderCenter {
  return FolioReaderCenter() // this is now public, so that users can return their subclass instead
}
```
This hides almost all of the internal state, but allows framework users to inject their own UI safely. I use the distinction between `create` and `setup` methods in my own code to make it very clear that `create` has side-effects (like changing the view hierarchy or global state), while `setup` simply sets up an object and returns it. I'm not too hung up on terminology, but I find that it lexically separates the code nicely.

In my case, I overrode `configureNavBarButtons()` on `FolioReaderCenter`, added my own bar button, and added a delegate to handle its action. The rest of the frameworks just chugs along! 
```swift
class MyCustomReaderContainer: FolioReaderContainer, MyCustomReaderCenterDelegate {

  weak var myCustomReaderContainerDelegate: MyCustomReaderCenterDelegate?

  override func setupCenterViewController() -> FolioReaderCenter {
    let center = MyCustomReaderCenter()
    center.myCustomReaderCenterDelegate = self
    return center
  }
}
```
And wham bam!

<img width="432" alt="screen shot 2016-05-31 at 11 58 44 am" src="https://cloud.githubusercontent.com/assets/903856/15686900/0d953722-2727-11e6-9845-d2e5bd113dfa.png">

**Comment by hebertialmeida** [comment]
I liked that, go ahead. You can simple make a pull request

**Comment by alexpopov** [comment]
Will do over the next couple days after work 😊

**Comment by hebertialmeida** [comment]
I am really glad that it helped you 😊 and thanks for contributing.

**Comment by danfsd** [comment]
Hey @alexpopov I've been going through your refactor code and I found it awesome. I just feel like there's yet little space for customizing the `FolioReaderContainer` (e.g: a container which doesn't have side panels or any other additional UI).

I think that this could be resolved by placing all this important code in a **base class** such as `FolioReaderBaseContainer`, and then placing all this additional **UI code** on the `FolioReaderContainer` which **extends** `FolioReaderBaseContainer`.

Additionally, making ebookDidLoad `public` should be good for the subclasses of `FolioReaderBaseContainer` to override it, like the `UIViewController` classes are able to do.

What are your thoughts on that?

**Comment by alexpopov** [comment]
I only just started refactoring; there is still a very long ways to go! (I haven't had much time to work on it lately)

I haven't looked through all the Ebook loading logic, but I feel like it should be factored out of the container (near) entirely so that it's divorced from the UI. Then `FolioReaderBaseContainer` could be a protocol, with one concrete implementation as currently provided by `FolioReaderContainer`, or any conforming class that the user chooses to implement. Having a *very* dumb `BaseContainer` class is also a very reasonable approach.

The major benefit to this is the separation of model and UI: there could be a ReaderCore, which is easily testable, stateful, and probably largely opaque, and then all the UI classes, whether the ones included in the project or ones that framework users create themselves, which just respond to changes in the model via delegation (such as your `ebookDidLoad`) 

tl;dr: I think that first we should just refactor existing UI classes to be publicly available and have a reasonable API—a very large task—and after that's done, opening up the possibility of using *any* UI, as you suggest, is a fantastic further step.

**Comment by hebertialmeida** [comment]
All classes are public, so closing this.
